twitchpopups.htm

```diff
@@ -102,24 +112,66 @@ <h1 id="popuptext"></h1>
-                  doAnimation();
+                    hotSeatIsOn = true;
+                    $("#popupbox").show();
+                    voteTopic = commandName.substr(10);
+                    $("#popupbox").css({"background-color":hotseatBg}); 
+                    $("#popuptext").text(`${hotseatEmoji}  ${hotSeatUser.toUpperCase()}`);
+                    doAnimation();
+                } else if (commandName.startsWith("!vote ")) {
+                    // Start with the topic for your vote
+                    $("#popupbox").show();
+                    voteTopic = commandName.substr(6).toUpperCase();
+                    $("#popupbox").css({"background-color": votingBg});
+                    $("#popuptext").text(`VOTE: ${voteTopic} 🔴`);
+                    doAnimation();
+                } else if (commandName.startsWith("!voteopt ")) {
+                    // Then add your options
+                    let option = commandName.substr(9).toUpperCase();
+                    options[option] = {};
+                } else if (commandName.startsWith("!voteoptclear")) {
+                    // Made a mistake? clear and re-add your options
+                    options = [];
+                } else if (commandName.startsWith("!votestart") && !votingIsOn) {
+                    // Once options are added, start the vote!
+                    votingIsOn = true;
+                    $("#popuptext").text(`VOTE: ${voteTopic} 🟢`);
+                } else if (commandName.startsWith("!voteresults") && votingIsOn) {
+                    // Display the results and stop the vote
+                    let biggestOption = null;
+                    let biggestSize = 0;
+                    for (let i = 0; i < Object.keys(options).length; i++) {
+                        if (Object.keys(options[Object.keys(options)[i]]).length > biggestSize) {
+                            biggestOption = Object.keys(options)[i];
+                            biggestSize = Object.keys(options[Object.keys(options)[i]]).length;
+                        } 
+                    }
+                    $("#popuptext").text(`AND THE WINNER IS: ${biggestOption} 👏👏👏`);
+                    doAnimation();
                 }
             }
-                
+
             if (hotSeatIsOn) {
                 if (context.username == hotSeatUser){
                     $("#popuptext").text(`${hotseatEmoji} ${context['display-name']}: ${commandName} ${hotseatEmoji}`);
                     doHotseatAnimation();
                 }
             }
                 
+            // If voting is on and the user typed somthing that can be a vote
+            if (votingIsOn && /^\d+$/.test(commandName) && parseInt(commandName) <= Object.keys(options).length) {
+                let vote = parseInt(commandName);
+                // Make sure users can only vote for one thing
+                for (let i = 0; i < Object.keys(options).length; i++) {

```

Comment:
Data structure suggestion: if you use a [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) of username → vote you don't have to do any work here:

```js
// let votes = new Map();
votes.set(context.username, vote);
```

Reply:
Yeah I like that idea, code simplicity > performance. I will work on changing this now.

Reply:
Just doing a bit of reading and it looks like map is not fully supported in IE: 
![image](https://user-images.githubusercontent.com/7206846/82292118-1f790d00-99a2-11ea-8ac5-64345f56ac68.png)

I wouldn't want a random streamer to download this and start to not work mid stream because they are in IE ¯\_(ツ)_/¯

Reply:
The OBS browser source uses the engine from Chrome.

Reply:
Another community member published an alternative simplification that gets around the map for now, thanks for the suggestion though!